File: Medium/Javascript/131. Palindrome Partitioning.js
/**Given a string s, partition s such that every 
substring
 of the partition is a 
palindrome
. Return all possible palindrome partitioning of s. */

/**
 * @param {string} s
 * @return {string[][]}
 */
var partition = function (s) {
  let res = [];
  let part = [];

  function dfs(i) {
    if (i >= s.length) {
      res.push(part.slice(0, part.length));
      return;
    }

    for (let j = i; j < s.length; j++) {
      if (checkPalindrome(s, i, j)) {
        part.push(s.slice(i, j + 1));
        dfs(j + 1);
        part.pop();
      }
    }
  }

  dfs(0)

  return res;
};

function checkPalindrome(s, i, j) {
  while (i <= j) {
    if (s[i] != s[j]) {
      return false;
    } else {
      i++;
      j--;
    }
  }
  return true;
}

console.log(partition("aba"))

File: Medium/python/131. Palindrome Partitioning.py
"""
Given a string s, partition s such that every substring of the partition is a palindrome. 
Return all possible palindrome partitioning of s.
"""

from typing import List


class Solution:
    def checkPali(self, s, i, j):
        l = i
        r = j
        while l < r:
            if s[l] != s[r]:
                return False
            else:
                l += 1
                r -= 1
        return True

    def partition(self, s: str) -> List[List[str]]:
        result = []
        part = []

        def dfs(i):
            if i >= len(s):
                result.append(part.copy())
                return

            for j in range(i, len(s)):
                if self.checkPali(s, i, j):
                    part.append(s[i: j+1])
                    dfs(j + 1)
                    part.pop()

        dfs(0)
        return result

print(Solution().partition("aba"))